source/change-streams/change-streams.rst

```diff
@@ -544,8 +525,34 @@ Option 1: ChangeStream::getResumeToken()
     public getResumeToken() Optional<Document>;
   }
 
-This method returns the cached ``documentResumeToken`` or ``postBatchResumeToken`` following rules from `Resume Process`_:
+This method returns the cached ``documentResumeToken`` or ``postBatchResumeToken`` following these rules:
+
+- If the ``ChangeStream`` has a cached ``postBatchResumeToken`` and has returned all documents in the most recent batch (or the most recent batch was empty), return the cached ``postBatchResumeToken`` and:
```

Comment:
Noting the tests that @DmitryLukyanov created in https://github.com/mongodb/specifications/pull/526/files, I think there are some additional edge cases to mention here:

 * `aggregate` has just executed. The returned batch is not empty and has not been iterated at all. We ignore any document resume token or `postBatchResumeToken` and only return a user-specified option (`startAfter` or `resumeAfter`) if available.
 * `getMore` has just executed. The returned batch is not empty and has not been iterated at all. We return the previous batch's `postBatchResumeToken` if available. Otherwise, return the last document resume token we saw in the previous batch. Otherwise, fall back to a user-specified option (`startAfter` or `resumeAfter`) if available.

Reply:
Something else that occurred to me while reading this PR and implementing `postBatchResumeToken` in libmongoc is that the spec is not entirely clear on _when_ to update the cached `postBatchResumeToken`.

[Driver API](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#driver-api) states the following for `postBatchResumeToken`:

> The most recent postBatchResumeToken returned in an aggregate or getMore response.

Later on in the spec, [ChangeStream](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#changestream) states:

> A change stream MUST track the latest postBatchResumeToken included in an aggregate or getMore response if available.

I think both of these can easily be misinterpreted, which may lead to bugs when implementing resume logic or `getResumeToken()`. The `getMore` case in my previous comment may need to return `postBatchResumeToken` from the _previous_ command response (`aggregate` or `getMore`) if the most recently executed `getMore` batch is non-empty and has not been iterated. If drivers decide to update their cached `postBatchResumeToken` value immediately upon receiving an `aggregate` or `getMore` response, they may lose that _previous_ value that may need to be returned before iterating a non-empty batch.

We'd do well to clarify exactly when the cached `postBatchResumeToken` should be updated.

 * If `getMore` returns a non-empty batch, we need to wait until at least one document has been iterated before overwriting the last-cached `postBatchResumeToken`.
 * If `getMore` returns an empty batch, there is no harm in updating `postBatchResumeToken` immediately.

I think the above clarification can go into this PR since it's closely related to resume tokens, but if anyone objects feel free to open a SPEC ticket to handle this separately.

Reply:
Good catch. I agree that's wrong or at least misleading.

I wonder if separating `postBatchResumeToken` and `documentResumeToken` in the [Driver API](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#driver-api) was a mistake. It makes the rules for resuming (well now `getResumeToken()`) simpler. Because, it lets us say "if x, return `postBatchResumeToken`, if y, return `documentResumeToken`". But if we need to add special rules of how we update the cached `postBatchResumeToken`, maybe we should go even further. Just have one `resumeToken` in `ChangeStream`. Something like:

```
interface ChangeStream extends Iterable<Document> {
  /**
   * A cached resume token.
   */
  private resumeToken: Document;
```

Then, we rework rules and put them in a new section ala:

**Updating the cached resume token**

- When the change stream is started:
    - If `resumeAfter` is specified, use that.
    - Otherwise, if `startAfter` is specified, use that.
    - Otherwise, `ChangeStream.resumeToken` remains unset.
- When `aggregate` or `getMore` returns:
    - If an empty batch was returned and a `postBatchResumeToken` is included, use that.
- When returning a document to the user:
    - If it's the last document in the batch and a `postBatchResumeToken` is included, use `postBatchResumeToken`.
    - Otherwise, use the `_id` of the document.

And then, `getResumeToken()` just returns `ChangeStream.resumeToken`.

Normally, I'd vote for creating a separate spec ticket instead of piling on more changes. But in this case we'll end up undoing the work here, so I vote for doing this as part of this PR.

@jmikola what do you think?

Reply:
"Updating the cached resume token" reads very clearly and would also greatly simplify logic for exposing the resume token (either getter or dispatching an event). I'd be in favor of this change.

Reply:
@kevinAlbs: One question. Would your initial case for "when the change stream is started" conflict with the following bullet in [Option 1: ChangeStream::getResumeToken()](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#option-1-changestream-getresumetoken)?

> If the ChangeStream has no cached documentResumeToken or postBatchResumeToken this returns an unset optional.

I assume we also wouldn't want to emit an event with the initial `resumeAfter` or `startAfter` option values, since users already have access to anything provided as a change stream option.

I do agree with consolidating the `documentResumeToken` and `postBatchResumeToken` tracking into a single field, though.

Reply:
@kevinAlbs: A correction to your last comment:

> **Updating the cached resume token**
> 
> - When the change stream is started:
>     - If `resumeAfter` is specified, use that.
>     - Otherwise, if `startAfter` is specified, use that.
>     - Otherwise, `ChangeStream.resumeToken` remains unset.

According to [Resume Process](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#resume-process), `startAfter` takes precedence over `resumeAfter`. I think those should be switched in your list.

Reply:
> One question. Would your initial case for "when the change stream is started" conflict with the following bullet in Option 1: ChangeStream::getResumeToken()?

Hah, yeah. I think `getResumeToken()` should have returned the value of `startAfter`/`resumeAfter` in the first place. If an application is saving the value of `getResumeToken()`, it should be able to start a new change stream with `resumeAfter: getResumeToken()` wherever it left off, even if nothing has been iterated. What I have in mind:

```
savedResumeToken = load("resumeToken.json")
cs = db.coll.watch({resumeAfter: savedResumeToken})
save(cs.getResumeToken(), "resumeToken.json")
/* application crashes here, before iteration. */
while (doc = cs.next()) {
    process(doc);
    save (cs.getResumeToken(), "resumeToken.json")
}
```

> I assume we also wouldn't want to emit an event with the initial resumeAfter or startAfter option values, since users already have access to anything provided as a change stream option.

By the same reasoning, I think we should actually emit an event.
```
savedResumeToken = load("resumeToken.json")
cs = db.coll.watch({resumeAfter: savedResumeToken})
cs.on("resumeTokenChanged", (token) => save(token, "resumeToken.json"))
/* application crashes before first document returned. */
cs.on("change", (doc) => { ... })
```

That keeps the logic consolidated. We have one "Updating the cached resume token" section. `getResumeToken()` returns that cached token. And when that cached resume token changes async drivers emit an event with that token.

> According to Resume Process, startAfter takes precedence over resumeAfter. I think those should be switched in your list.

Order shouldn't matter. startAfter and resumeAfter are mutually exclusive.

Reply:
> I think we should actually emit an event.

In the event-emitting example you provided, I suppose the `resumeTokenChanged` handler isn't registered until _after_ the change stream is constructed with `watch()` so it may never need to receive a redundant event for the initial option as that would have occurred _during_ `watch()`.

> Order shouldn't matter. startAfter and resumeAfter are mutually exclusive.

Drivers rely on the server to raise an error if both are specified, but there are several libmongoc tests where we specify both at once and assert what happens with the outgoing command. I foresee doing the same for `getResumeToken()` tests, so this could come up if we're using a mock server.

Reply:
> In the event-emitting example you provided, I suppose the resumeTokenChanged handler isn't registered until after the change stream is constructed with watch() so it may never need to receive a redundant event for the initial option as that would have occurred during watch().

I dunno, I guess that depends on implementation. Like, right now in Node attaching a "change" event listener [triggers the change stream iteration](http://mongodb.github.io/node-mongodb-native/3.2/api/ChangeStream.html):

> Attaching a change event listener to a Change Stream will switch the stream into flowing mode. Data will then be passed as soon as it is available.

Plus, it might not be redundant. Maybe a more compelling example is when you get a change stream already constructed, and don't have the initial options accessible:

```
cs = getSomeChangeStreamObject();
cs.on("resumeTokenChanged", (token) => save(token, "resumeToken.json"))
```

Foreseeably, a "resumeTokenChanged" event listener could be called immediately with the value of `startAfter`/`resumeAfter` if the change stream hasn't started iterating internally. But then, I guess there'd be weirdness of different not getting the first "resumeTokenChanged" event if you attach a "change" event first...

I vote we leave this up to the implementation and stay vague, like we have it now: "Allow users to set a callback to listen for new resume tokens.". If you think we should make a note of it, let's not continue piling on in this PR, and just create another spec ticket to clarify the first token that should be emitted in "resumeTokenChanged".

> Drivers rely on the server to raise an error if both are specified, but there are several libmongoc tests where we specify both at once and assert what happens with the outgoing command. I foresee doing the same for getResumeToken() tests, so this could come up if we're using a mock server.

Fair enough. I guess we need to stick to my original arbitrary ordering `startAfter` takes precedence over `resumeAfter`.